### PR TITLE
Health and stamina

### PR DESCRIPTION
**Pull request by 0xNonCents**
Stamina 
Idle, walking and blocking nets the character stamina. Each frame in these actions provides stamina.
Attacking and Dashing cost stamina upfront in order to execute the move. If a character does not have sufficient stamina to perform what it intends then the character either continues the current action or returns to IDLE.

Max Stamina : 1000

Integrity
Being Hurt or Knocked costs the character 100 integrity. 

Max Integrity : 1000

Screen Shot 2023-02-10 at 11 38 33 AM

----
What happens when fatigued?

Cancellable moves (not last frame) - 
 -  increment the counter
 - indicate fatigue

Last Frame (with ability to act again) -
- go to IDLE
- indicate fatigue

During actions that recover stamina -
- increment counter/continue action
- indicate fatigue

### CODE REVIEW

File: 8-zed/contracts/body/body_antoc.cairo [hunk 1]
```diff
@@ -89,12 +101,13 @@ func _body_antoc {range_check_ptr}(
...
+            if (intent == ns_antoc_action.HORI ) {
                 // return to first frame
-                return ( body_state_nxt = BodyState(ns_antoc_body_state.HORI, 0, integrity, stamina, dir) );
-            } else {
+                if(enough_stamina == 1){
+                    return ( body_state_nxt = BodyState(ns_antoc_body_state.HORI, 0, integrity, updated_stamina, dir) );
+                }
                 // otherwise return to IDLE
-                return ( body_state_nxt = BodyState(ns_antoc_body_state.IDLE, 0, integrity, stamina, dir) );
+                return ( body_state_nxt = BodyState(ns_antoc_body_state.IDLE, 0, integrity, default_stamina, dir) );
```
Comment (0xNonCents) [inline 8-zed/contracts/body/body_antoc.cairo:110]: on the frame a character is returned to idle by force (not enough stamina to perform intent, recovering from hurt/knock animation) they are given the stamina reward equivalent to IDLE.

File: 8-zed/contracts/body/body_antoc.cairo [hunk 2]
```diff
@@ -124,18 +137,21 @@ func _body_antoc {range_check_ptr}(
...
                 // return to first frame
-                return ( body_state_nxt = BodyState(ns_antoc_body_state.VERT, 0, integrity, stamina, dir) );
-            } else {
+                if(enough_stamina == 1){
+                    return ( body_state_nxt = BodyState(ns_antoc_body_state.VERT, 0, integrity, updated_stamina, dir) );
+                }
                 // otherwise return to IDLE
-                return ( body_state_nxt = BodyState(ns_antoc_body_state.IDLE, 0, integrity, stamina, dir) );
+                return ( body_state_nxt = BodyState(ns_antoc_body_state.IDLE, 0, integrity, default_stamina, dir) );
             }
```
Comment (0xNonCents) [inline 8-zed/contracts/body/body_antoc.cairo:155]: In the diff it is hard to see but I am using short circuit evaluation to reduce the number of if-elses in the code.

File: 8-zed/contracts/body/body_antoc.cairo [hunk 3]
```diff
@@ -26,47 +30,55 @@ func _body_antoc {range_check_ptr}(
...
         }
 
         // body responds to intent; locomotive action has lowest priority
-        if (intent == ns_antoc_action.HORI) {
-            return ( body_state_nxt = BodyState(ns_antoc_body_state.HORI, 0, integrity, stamina, dir) );
-        }
-        if (intent == ns_antoc_action.VERT) {
-            return ( body_state_nxt = BodyState(ns_antoc_body_state.VERT, 0, integrity, stamina, dir) );
-        }
         if (intent == ns_antoc_action.BLOCK) {
```
Comment (0xNonCents) [inline 8-zed/contracts/body/body_antoc.cairo:52]: I moved BLOCK above the attack moves in order to enclose all of the intents that cost stamina in one
```
if (enough_stamina == 1){
...actions that cost stamina
}
```

**Comment by 0xNonCents** [comment]
BIG TODO - Integrity losses from CLASHING

**Comment by 0xNonCents** [comment]
Other things on my mind but not blockers:
- Is the idle frame that comes after an action a part of the action? (does not get to recover stamina during it)
- Balancing costs and recovery speeds. 
- should each frame of an action such as block recover a part of the stamina recovery? (5 frames, +25 stamina => 5 stamina each frame)

**Comment by guiltygyoza** [comment]
> should each frame of an action such as block recover a part of the stamina recovery? (5 frames, +25 stamina => 5 stamina each frame)

I think for recovery it makes total sense. For consumption, we can also do that - amortizing consumption across frames - and if stamina runs out midway, body transitions to some tiring state. But we would need excellent animation support in this case, otherwise the transition from action to tiring would look very glitchy.

**Comment by guiltygyoza** [comment]
would appreciate writing a short summary for the PR up top

File: 8-zed/contracts/body/body_antoc.cairo (hunk 1 above)
Comment (guiltygyoza) [reply 8-zed/contracts/body/body_antoc.cairo]: scary - I'm not seeing this diff when pulling to my local. What is `default_stamina`?

**Comment by guiltygyoza** [comment]
Questions:
1. What purpose does `calculate_default_stamina()` serve? Its return value is currently not used.
2. It would be useful to add the `else` branch for `if(enough_stamina == TRUE){...}` and set some binary flag in the Frame. This would allow the frontend to visualize the effect "agent's Mind attempted an intent but Body has insufficient stamina to carry out the intent".  We can implement this in a separate PR.

**Comment by 0xNonCents** [comment]
Adding the fatigue indicator now

File: 8-zed/contracts/constants/constants.cairo [hunk 4]
```diff
@@ -22,6 +22,20 @@ namespace ns_stamina {
     const INIT_STAMINA = 1000;
 }
 
+
+namespace ns_common_stamina_effect {
```
Comment (greged93) [inline 8-zed/contracts/constants/constants.cairo:26]: Suggest updating the constant values to negative values if removing stamina and positive values if adding stamina.

File: 8-zed/contracts/body/body_utils.cairo [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+    
+    with_attr error_message("Intent is not recognized.") {
+        assert 0 = 1;
+    }
+    return (0, FALSE);
+}
+
+func _settle_stamina_change{range_check_ptr}(stamina : felt, stamina_change : felt, max_stamina : felt, is_add : felt) -> (felt, felt) {
```
Comment (greged93) [inline 8-zed/contracts/body/body_utils.cairo:79]: Suggest updating the this function and follow the pattern used in [dynamics](https://github.com/topology-gg/shoshin/blob/main/8-zed/contracts/dynamics.cairo#L255). This allows us to have positive values for increase of stamina and negative values for decrease of stamina, as well as remove the `is_add` parameter.

File: 8-zed/contracts/body/body_antoc.cairo [hunk 6]
```diff
@@ -278,45 +309,48 @@ func _body_antoc {range_check_ptr}(
...
+            if (intent == ns_antoc_action.DASH_BACKWARD) {            
+                return ( body_state_nxt = BodyState(ns_antoc_body_state.DASH_BACKWARD, 0, integrity, updated_stamina, dir, FALSE) );            
+            }
         }
 
         // continue moving forward
         if (intent == ns_antoc_action.MOVE_BACKWARD) {
             if (counter == ns_antoc_body_state_duration.MOVE_BACKWARD - 1) {
-                return ( body_state_nxt = BodyState(ns_antoc_body_state.MOVE_BACKWARD, 0, integrity, stamina, dir) );
+                return ( body_state_nxt = BodyState(ns_antoc_body_state.MOVE_BACKWARD, 0, integrity, updated_stamina, dir, is_fatigued) );
```
Comment (greged93) [inline 8-zed/contracts/body/body_antoc.cairo:340]: Here you return `is_fatigued` whereas in [`state == ns_antoc_body_state.BLOCK`](https://github.com/topology-gg/shoshin/blob/6347cc94c43f234d8704f599a18d702261492d61/8-zed/contracts/body/body_antoc.cairo#L192) you return `FALSE`, is there a reason for the difference? From your diagram I understand that `BLOCK`, `MOVE_BACKWORD` and `MOVE_FORWARD` work in a similar fashion regarding stamina.

**Comment by greged93** [comment]
Also @0xNonCents, would it be possible to add a test of your functions? We need to add more tests into the code, maybe even start unit testing everything (which is a big task since little was done up to now).

File: 8-zed/contracts/engine.cairo [hunk 7]
```diff
@@ -213,6 +213,23 @@ func _loop{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
     // Preparation
     //
     let last_frame: FrameScene = arr_frames[idx - 1];
+    
+
+    //
+    /// See if a player has health <= 0
+    //
+
+    let agent_0_standing = is_in_range(last_frame.agent_0.body_state.integrity, 0, 101);
```
Comment (greged93) [inline 8-zed/contracts/engine.cairo:222]: Should be 1001 since INIT_INTEGRITY is 1001. Propose to change to INIT_INTEGRITY + 1. Same on the next line.

Comment (0xNonCents) [reply 8-zed/contracts/engine.cairo:222]: yes. good find

File: 8-zed/contracts/body/body_antoc.cairo (hunk 6 above)
Comment (0xNonCents) [reply 8-zed/contracts/body/body_antoc.cairo]: Good find. I changed it from is_fatigued to FALSE for quite a few of these body_states

**Comment by greged93** [comment]
lgtm !

**Review by greged93 — APPROVED** [review]

**Review by guiltygyoza — APPROVED** [review]
love the unit tests and your thoughtfulness in how stamina consumption is handled

**Pull request merged**